### PR TITLE
Standardise newlines after module-level docstrings

### LINKED ISSUE #1872: Enforce blank line after module docstring
Black does not have an opinion on whether there should be a blank line immediately after a module's docstring. This causes inconsistencies across blackened files because whatever the author chooses, stays.

Both examples below satisfy black:
```python
"""Module docstring."""

import typing
```

```python
"""Module docstring."""
import typing
```

My preference would be to enforce a blank line.

Related issue: https://github.com/psf/black/issues/450

### PR DESCRIPTION
**Pull request by jpy-git**
### Description

Closes #1872.

Does 2 things:
a) Refactor `EmptyLineTracker` for more consistent look-back logic. Simplifies logic by removing complex interactions between `before` and `after` for various rules.

b) Standardise newlines after module-level docstrings. We now enforce 1 newline after both single and multi line docstrings.
e.g.
```python
"""docstring"""

import foo
```
```python
"""docstring

blah blah blah
"""

import foo
```

### Checklist - did you ...

- [X] Add a CHANGELOG entry if necessary?
- [X] Add / update tests if necessary?
- [X] Add new / update outdated documentation?

### CODE REVIEW

**Comment by jpy-git** [comment]
This is diff-shades with no newline after multi-line docstrings
```
╭───────────────────────── Summary ─────────────────────────╮
│ 15 projects & 351 files changed / 366 changes [+115/-251] │
│                                                           │
│ ... out of 2 177 706 lines, 10 484 files & 23 projects    │
╰───────────────────────────────────────────────────────────╯

Differences found.
```
Will make another commit now with 1 newline after multi-line docstrings to help make a decision on which is best

File: src/black/lines.py [hunk 1]
```diff
@@ -419,37 +419,47 @@ def __bool__(self) -> bool:
...
+        if (
+            self.preview
+            and len(self.previous_lines) == 1
+            and self.previous_lines[-1].is_triple_quoted_string
+        ):
+            # Newlines after multi-line module level docstring.
+            if str(self.previous_lines[-1].leaves[0]).count("\n") >= 1:
+                before = 0
+            else:
+                before = 1
```
Comment (jpy-git) [inline src/black/lines.py:457]: This is the only new logic

**Comment by jpy-git** [comment]
Intuitively that doesn't feel unreasonable to me

Looking at diff shades it only runs on 23 projects which isn't a large sample:
https://github.com/ichard26/diff-shades/blob/6ec675b686963e5ccb21f562cc1c04f48c23eb10/src/diff_shades/config.py#L35-L72

The 10k files number will be very heavily inflated by the few large projects on that list rather than being evenly distributed amongst those 23 projects. e.g. `django` has 2743 python files, whereas `aioexagb` only has 20 python files.

In my experience multi-line module level docstrings are relatively uncommon as people tend to only use them in files that are part of a public API and that a user would realistically call `help` on, and not in every internal file where they're much more likely to use a single-line or just omit the docstring completely. If you randomly look through files in Django you can see most don't have any docstring.

If you look at the current output of diff shades, you can see that all the changed lines are acting exactly as we would expect from this change.

^Will whip up a quick script to scrape these projects and verify these points above 😄

**Comment by jpy-git** [comment]
Quick and dirty script to categorise files in each of the projects into `no_docstring`, `single_line_docstring`, and `multi_line_docstring`:
```python
from collections import defaultdict
import json
import git
import os
import re
import shutil
import tempfile

PROJECTS = [
    ("aioexabgp", "https://github.com/cooperlees/aioexabgp.git"),
    ("attrs", "https://github.com/python-attrs/attrs.git"),
    ("bandersnatch", "https://github.com/pypa/bandersnatch.git"),
    ("black", "https://github.com/psf/black.git"),
    ("blackbench", "https://github.com/ichard26/blackbench.git"),
    ("channel", "https://github.com/django/channels.git"),
    ("diff-shades", "https://github.com/ichard26/diff-shades.git"),
    ("django", "https://github.com/django/django.git"),
    ("flake8-bugbear", "https://github.com/PyCQA/flake8-bugbear.git"),
    ("hypothesis", "https://github.com/HypothesisWorks/hypothesis.git"),
    ("pandas", "https://github.com/pandas-dev/pandas.git"),
    ("pillow", "https://github.com/python-pillow/Pillow.git"),
    ("poetry", "https://github.com/python-poetry/poetry.git"),
    ("ptr", "https://github.com/facebookincubator/ptr.git"),
    ("pyanalyze", "https://github.com/quora/pyanalyze.git"),
    ("pyramid", "https://github.com/Pylons/pyramid.git"),
    ("pytest", "https://github.com/pytest-dev/pytest.git"),
    ("scikit-lego", "https://github.com/koaning/scikit-lego.git"),
    ("sqlalchemy", "https://github.com/sqlalchemy/sqlalchemy.git"),
    ("tox", "https://github.com/tox-dev/tox.git"),
    ("typeshed", "https://github.com/python/typeshed.git"),
    ("virtualenv", "https://github.com/pypa/virtualenv.git"),
    ("warehouse", "https://github.com/pypa/warehouse.git"),
]


project_counts = {}
tmp_dir = tempfile.mkdtemp()
for name, url in PROJECTS:
    docstring_counts = defaultdict(int)
    tmp_path = os.path.join(tmp_dir, name)
    git.Repo.clone_from(url, tmp_path)
    # Only want to count files on default branch
    shutil.rmtree(os.path.join(tmp_path, ".git"))
    for dir, _, files in os.walk(tmp_path):
        for file in files:
            if not file.endswith(".py"):
                continue
            
            with open(os.path.join(dir, file), "r") as f:
                code = f.read()
            
            # Try and match docstring from start of file
            match = re.match(r'""".*?"""', code, re.DOTALL)

            # Catagorise files
            if match is None:
                docstring_counts["no_docstring"] += 1
            elif match.group(0).count("\n") >= 1:
                docstring_counts["multi_line_docstring"] += 1
            else:
                docstring_counts["single_line_docstring"] += 1
    
    project_counts[name] = docstring_counts


print(json.dumps(project_counts, indent=4))
```
Results
```
{
    "aioexabgp": {
        "no_docstring": 20
    },
    "attrs": {
        "no_docstring": 49
    },
    "bandersnatch": {
        "no_docstring": 49,
        "multi_line_docstring": 3
    },
    "black": {
        "no_docstring": 140,
        "multi_line_docstring": 15,
        "single_line_docstring": 6
    },
    "blackbench": {
        "no_docstring": 29,
        "single_line_docstring": 2,
        "multi_line_docstring": 8
    },
    "channel": {
        "no_docstring": 44
    },
    "diff-shades": {
        "no_docstring": 14,
        "multi_line_docstring": 1
    },
    "django": {
        "no_docstring": 2504,
        "multi_line_docstring": 219,
        "single_line_docstring": 20
    },
    "flake8-bugbear": {
        "no_docstring": 14,
        "multi_line_docstring": 18
    },
    "hypothesis": {
        "no_docstring": 403
    },
    "pandas": {
        "no_docstring": 1072,
        "multi_line_docstring": 279,
        "single_line_docstring": 61
    },
    "pillow": {
        "no_docstring": 265,
        "single_line_docstring": 1,
        "multi_line_docstring": 10
    },
    "poetry": {
        "multi_line_docstring": 3,
        "no_docstring": 304
    },
    "ptr": {
        "no_docstring": 5
    },
    "pyanalyze": {
        "no_docstring": 38,
        "multi_line_docstring": 34
    },
    "pyramid": {
        "no_docstring": 544,
        "single_line_docstring": 10,
        "multi_line_docstring": 1
    },
    "pytest": {
        "no_docstring": 189,
        "multi_line_docstring": 19,
        "single_line_docstring": 36
    },
    "scikit-lego": {
        "no_docstring": 108,
        "multi_line_docstring": 1,
        "single_line_docstring": 4
    },
    "sqlalchemy": {
        "multi_line_docstring": 83,
        "no_docstring": 510,
        "single_line_docstring": 22
    },
    "tox": {
        "no_docstring": 79,
        "single_line_docstring": 9,
        "multi_line_docstring": 5
    },
    "typeshed": {
        "no_docstring": 10
    },
    "virtualenv": {
        "no_docstring": 133,
        "multi_line_docstring": 9,
        "single_line_docstring": 9
    },
    "warehouse": {
        "no_docstring": 528
    }
}
```

**Comment by jpy-git** [comment]
TLDR: As expected total file count is heavily skewed by `django`, `pandas`, `sqlalchemy`, and `warehouse`.

Total counts:
no_docstring: 7051
single_line_docstring: 180
multi_line_docstring: 708

as expected most files have no docstring (although surprised that not many use single lines).

That `multi_line_docstring` number reflects the results we saw from diff-shades [here](https://github.com/psf/black/pull/2996#issuecomment-1089460414) so think it all looks reasonable 😄

**Review by felix-hilden — COMMENTED** [review]
Man, huge thanks for taking on all these issues!

I think I'd prefer the newlines to be added to multiline docstrings as well, just for consistency. Not strongly though, and against my own habits and what I said in the original issue I know, but still.

About the preview handling: certainly this is a feature we'd carry over without question, so no biggie, but having the clarity of being able to search for usages of a potential `Preview.docstring_newlines` (or similar) was intended to help in picking individual changes for stable. Not that it's likely that we have other changes in `EmptyLineTracker` so it would be fairly easy to find this one thing, but anyways. Example PR: #2916

I'm trying to slowly get into the processing code. Your modifications seem sensible, but I'm not the authority on that I'm afraid 😄 I do trust the tests though, nice work!

**Comment by jpy-git** [comment]
> About the preview handling: certainly this is a feature we'd carry over without question, so no biggie, but having the clarity of being able to search for usages of a potential `Preview.docstring_newlines` (or similar) was intended to help in picking individual changes for stable.

Good point, will update to reference the specific preview mode 😄 


> I think I'd prefer the newlines to be added to multiline docstrings as well, just for consistency.

Yeah I'm easy either way, the logic for newlines after multiline docstrings is slightly simpler and consistency is always nice!

@JelleZijlstra @ichard26 Do you guys have any preference for no newline vs newline after multi-line module-level docstrings?

**Review by felix-hilden — COMMENTED** [review]
Much better, and a nice simplification! Again, provided that other maintainers also agree with the style, and that the implementation itself is sound.

**Comment by JelleZijlstra** [comment]
Unfortunately this has a lot of merge conflicts, sorry for dropping the ball here.

**Comment by aqeelat** [comment]
@JelleZijlstra does it mean that you need help rebasing it?

**Comment by JelleZijlstra** [comment]
Yes, somebody will have to rebase this.

**Comment by DanielNoord** [comment]
@JelleZijlstra I can't rebase on this PR but I'd be happy to invest the time to get this finished. Are you okay with me rebasing this and opening a new PR? (Or somebody else for that matter)

**Comment by DanielNoord** [comment]
Went ahead and created https://github.com/psf/black/pull/3932, please consider reviewing it. Very much open to feedback on how to change the code to make the PR mergeable!

**Comment by JelleZijlstra** [comment]
Merged #3932, thanks!
